src/librustc_mir/monomorphize/collector.rs

```diff
@@ -864,12 +865,14 @@ fn create_mono_items_for_vtable_methods<'tcx>(
     impl_ty: Ty<'tcx>,
     output: &mut Vec<MonoItem<'tcx>>,
 ) {
-    assert!(
-        !trait_ty.needs_subst()
-            && !trait_ty.has_escaping_bound_vars()
-            && !impl_ty.needs_subst()
-            && !impl_ty.has_escaping_bound_vars()
-    );
+    if !tcx.sess.opts.debugging_opts.polymorphize {
+        assert!(
+            !trait_ty.needs_subst()
+                && !trait_ty.has_escaping_bound_vars()
+                && !impl_ty.needs_subst()
+                && !impl_ty.has_escaping_bound_vars()
```

Comment:
The `has_escaping_bound_vars` checks should be unconditional.

Reply:
Fixed in 5b3c02b.